client/src/controllers/TabsController.test.js

```diff
@@ -0,0 +1,256 @@
+import { Application } from '@hotwired/stimulus';
+import { TabsController } from './TabsController';
+
+function simulateKeydown(key) {
+  const event = new KeyboardEvent('keydown', {
+    key,
+    code: key,
+    bubbles: true,
+    cancelable: true,
+  });
+  document.activeElement.dispatchEvent(event);
+}
+
+jest.useFakeTimers();
+
+describe('TabsController', () => {
+  let app;
+  const oldWindowLocation = window.location;
+
+  const setup = async (html) => {
+    document.body.innerHTML = `<main>${html}</main>`;
+
+    app = Application.start();
+    app.register('w-tabs', TabsController);
+    await Promise.resolve();
+  };
+
+  beforeAll(() => {
+    delete window.location;
+
+    window.location = Object.defineProperties(
+      {},
+      {
+        ...Object.getOwnPropertyDescriptors(oldWindowLocation),
+        assign: { configurable: true, value: jest.fn() },
+      },
+    );
+  });
+
+  afterEach(() => {
+    app?.stop();
+    jest.clearAllMocks();
+  });
+
```

Comment:
I think these tests are looking good but need a bit of a re-org.

Think about `describe` as a group and `it` as a test within that group.

Each `it` should build a brand new element on its own.

e.g.
```js
    it('should have aria-selected set to the first tab on connect', async () => {
      await setup();
     // rest of the test
```

Instead of adding nested sets of `beforeEach`.

While it's OK the way you have it it will be easier to add more tests with each `it` block doing it's own setup.

Can you also re-group the current tests as follows (or similar), the idea being you have a basic set of tests in one describe block. Then add describe blocks as you test different groups of behaviour.


1. describe('basic behavior'
  - Initial load (this test should check that `w-tabs:ready` has been dispatched - needs to be added to the controller
  - Selecting a tab with click (this test should also check for the classes in one it block)
  - Selecting a tab with keyboard
2. describe('loading with an activated tab
  - Loading with a different selected tab (this would have the different HTML with selected value changed.
  - Loading with a tab selected by the URL
3. describe('error handling
  - Validating of malformed aria attributes